fix: added field validation helper for aggregations

mdashti opened the pull request:
# Ticket(s) Closed

- Closes #2767

## What

Adds a public `validate_aggregation_fields()` helper function that validates all fields in an aggregation request exist in the schema and are configured as fast fields. This allows users to validate aggregations upfront before execution.

## Why

Users were getting confusing empty results when aggregation field names had typos or didn't exist. However, some users (like Quickwit) need lenient behavior for schema evolution across different data segments.

This PR provides a validation helper that users can call explicitly when they want validation, while keeping the default aggregation behavior lenient for backward compatibility.

## How

Added `validate_aggregation_fields()` function in `agg_req.rs` that:
- Extracts all field names from the aggregation request using existing `get_fast_field_names()`
- Validates each field exists in the schema (directly or as part of a JSON field)
- Checks that each field is configured as a fast field
- Returns `FieldNotFound` or `SchemaError` if validation fails

Users can call this before executing aggregations:

```rust
// Validate upfront (optional)
validate_aggregation_fields(&agg_req, segment_reader)?;

// Execute aggregation
let collector = AggregationCollector::from_aggs(agg_req, Default::default());
let results = searcher.search(&query, &collector)?;
```

The validation is completely optional - existing code continues to work unchanged with lenient behavior.

## Tests

- Added `test_aggregation_field_validation_helper` testing both invalid and valid fields
- Includes working doctest example
- All 166 existing aggregation tests pass unchanged
- Fully backward compatible - no breaking changes

mdashti committed:
Fixed agg validation

fulmicoton commented:
I thought I commented here, but can't find my comment :(.
Not sure what happened.

In quickwit, we make it possible to change your schema.
In that case, we check that the request for the current version of the schema, but we run the query over older data, with a best effort approach.

If the fields are not fast fields, the aggregation result will be empty.

So this is not a bug, but a feature.
Can we make your validation optional (and not the default)?

mdashti committed:
Merge remote-tracking branch 'tantivy/main' into paradedb/fix-agg-validation

mdashti committed:
Added a flag: strict_validation

fulmicoton-dd commented on src/aggregation/accessor_helpers.rs:
```diff
@@ -89,16 +115,39 @@ pub(crate) fn get_dynamic_columns(
     if ff_field_with_type.is_empty() {
+        if strict_validation {
+            // Check if the field exists in the schema but is not a fast field
+            let schema = reader.schema();
```
that's the same code as above isn't it?  Can you make a function somewhere.

fulmicoton-dd commented:
I added a few comments @mdashti. 

The PR contains what looks like AI slop.
Using AI agent is ok, but make sure you do not push the burden of CR on us.

mdashti committed:
Applied PR comments.

mdashti committed:
Revert "Revert all impl."

This reverts commit d5e0991549a05bf80f19f853f7689ad69f96e7e5.

mdashti committed:
Applied PR comment:
I would move it outside of the aggregation. You can fetch the fields from the aggregation request and do a validation in a helper function

mdashti commented on src/aggregation/agg_req.rs:
```diff
@@ -115,6 +115,54 @@ pub fn get_fast_field_names(aggs: &Aggregations) -> HashSet<String> {
     fast_field_names
 }
 
+/// Validates that all fields referenced in the aggregation request exist in the schema
```
a

mdashti commented on src/aggregation/accessor_helpers.rs:
```diff
@@ -55,22 +55,48 @@ pub(crate) fn get_numeric_or_date_column_types() -> &'static [ColumnType] {
+        Some(field) => Ok(field),
+        None => {
+            if strict_validation {
+                // Check if the field exists in the schema but is not a fast field
```
Applied in [90ceec6](https://github.com/quickwit-oss/tantivy/pull/2768/commits/90ceec62aa9832de5f2daaf042d95aab97f120c4), but then changed the approach based on your last comment.

mdashti committed:
Fixed the example

mdashti reviewed (commented):
@fulmicoton Thanks for the comments. Your idea of providing a helper function is a good one. Changed the approach. Please take another look.

mdashti committed:
Rename validate_aggregation_fields to validate_aggregation_fields_exist

Applies @PSeitz's review suggestion to make the function name more
descriptive of what it checks. Also adds a doc note clarifying why
validation is opt-in rather than enforced by default.

mdashti reviewed (commented):
@PSeitz Thanks for the comment. Please take another look.
